{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import pystan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '../../data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "pth = os.path.join(data_dir, 'pptc-pdx-clinical-web.txt')\n",
    "\n",
    "pdx_clinical = pd.read_csv(pth, sep='\\t')\n",
    "\n",
    "# Only consider models that are part of the PPTC\n",
    "pdx_clinical = pdx_clinical[pdx_clinical['RNA.Part.of.PPTC'] == 'yes']\n",
    "\n",
    "# Fix for the latest round of the paper\n",
    "assert pdx_clinical.loc[pdx_clinical['Model'] == 'ALL-102', 'Histology.Detailed'].item() == 'BCP-ALL'\n",
    "pdx_clinical.loc[pdx_clinical['Model'] == 'ALL-102', 'Histology.Detailed'] = 'Ph-likeALL'\n",
    "\n",
    "assert pdx_clinical.loc[pdx_clinical['Model'] == 'ALL-105', 'Histology.Detailed'].item() == 'BCP-ALL'\n",
    "pdx_clinical.loc[pdx_clinical['Model'] == 'ALL-105', 'Histology.Detailed'] = 'Ph-likeALL'\n",
    "\n",
    "assert pdx_clinical.loc[pdx_clinical['Model'] == 'ALL-115', 'Histology.Detailed'].item() == 'BCP-ALL'\n",
    "pdx_clinical.loc[pdx_clinical['Model'] == 'ALL-115', 'Histology.Detailed'] = 'Ph-likeALL'\n",
    "\n",
    "assert pdx_clinical.loc[pdx_clinical['Model'] == 'PAKYEP', 'Histology.Detailed'].item() == 'BCP-ALL'\n",
    "pdx_clinical.loc[pdx_clinical['Model'] == 'PAKYEP', 'Histology.Detailed'] = 'Ph-likeALL'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Attributes File for TumorMap Analysis\n",
    "\n",
    "pth = os.path.join(data_dir, 'tumormap-attr-2019-07-30-pdx.tsv')\n",
    "\n",
    "attr = pd.DataFrame(columns=['Model', 'Histology'])\n",
    "\n",
    "for (model, hist), rows in pdx_clinical.groupby(['Model', 'Histology.Detailed']):\n",
    "    \n",
    "    attr.loc[len(attr), :] = [model, hist]\n",
    "    \n",
    "attr.to_csv(pth, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load PDX expression data normalized to TPM\n",
    "pth = os.path.join(data_dir, 'pdx-TPM-2019-02-15.tsv')\n",
    "exp = pd.read_csv(pth, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started with  244\n",
      "Ended with  244\n"
     ]
    }
   ],
   "source": [
    "# Check how many of the PDXs have expression data\n",
    "\n",
    "pdxs = pdx_clinical['Model'].values\n",
    "print('Started with ', len(pdxs))\n",
    "\n",
    "for pdx in pdxs:\n",
    "    if pdx not in exp.columns:\n",
    "        print(pdx)\n",
    "\n",
    "pdxs = [x for x in pdxs if x in exp.columns]\n",
    "print('Ended with ', len(pdxs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove genes that have more than 80% of the genes non-expressed\n",
    "# Remove genes that are in the bottom 20% for low variance\n",
    "def expression_variance_filter(norm_df):\n",
    "    \"\"\"\n",
    "    This function was taken from the UCSC Treehouse protocol pipeline:\n",
    "    https://github.com/UCSC-Treehouse/protocol/blob/master/3_generate-thresholds.ipynb\n",
    "    \"\"\"\n",
    "    proportion_unexpressed = 0.8\n",
    "    variance_filter_level = 0.2\n",
    "\n",
    "    max_ok_zeroes = len(norm_df.columns) * proportion_unexpressed\n",
    "    def sufficiently_expressed(series):\n",
    "        return len(series[series <= 0.]) < max_ok_zeroes\n",
    "\n",
    "    withZeroes = norm_df.apply(sufficiently_expressed, axis=1) # Series: Gene to True (keep) or False\n",
    "    expression_filtered_df = norm_df[withZeroes]\n",
    "\n",
    "    variance = expression_filtered_df.apply(np.std, axis=1) #  Series\n",
    "    cut_proportion = int(math.ceil(len(variance) * variance_filter_level))\n",
    "    keep_proportion = len(variance) - cut_proportion\n",
    "\n",
    "    # list of genes that remain after\n",
    "    expr_var_filtered_genelist = variance.nlargest(keep_proportion) #  Series\n",
    "\n",
    "    return expr_var_filtered_genelist.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset expression matrix to the PPTC PDXs\n",
    "exp = exp[pdxs]\n",
    "\n",
    "# Apply log2(TPM + 1) transformation\n",
    "exp = np.log2(exp + 1.0)\n",
    "\n",
    "# Filter genes that are mostly zero or have \n",
    "# variance below the 20th percentile\n",
    "filt_genes = expression_variance_filter(exp)\n",
    "exp = exp.reindex(filt_genes)\n",
    "\n",
    "# Center data\n",
    "exp = exp.apply(lambda x: x - x.mean(), axis=1) \n",
    "\n",
    "# Drop missing values\n",
    "exp.dropna(inplace=True)\n",
    "\n",
    "exp.shape\n",
    "\n",
    "pth = os.path.join(data_dir, 'tumormap-exp-2019-07-30-pdx.tsv')\n",
    "exp.to_csv(pth, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Phase</th>\n",
       "      <th>PersonID</th>\n",
       "      <th>DNA.Part.of.PPTC</th>\n",
       "      <th>RNA.Part.of.PPTC</th>\n",
       "      <th>Have.snp.file</th>\n",
       "      <th>snp.array.filename</th>\n",
       "      <th>snp.array.sample.ID</th>\n",
       "      <th>Have.maf</th>\n",
       "      <th>WES.human.bam.filename</th>\n",
       "      <th>...</th>\n",
       "      <th>Histology.Oncoprints2</th>\n",
       "      <th>STR.OK</th>\n",
       "      <th>Model.Type</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Histology.Detailed</th>\n",
       "      <th>PI</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Reported_Ethnicity</th>\n",
       "      <th>Inferred_Ethnicity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Histology.Priors</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BCP-ALL</th>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Brain</th>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>48</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>42</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Carcinoma</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETP-ALL</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ewing Sarcoma</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Extracranial Rhabdoid</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MLL-ALL</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neuroblastoma</th>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>28</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Osteosarcoma</th>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other Sarcoma</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ph+/like-ALL</th>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Renal</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rhabdomyosarcoma</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-ALL</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Model  Phase  PersonID  DNA.Part.of.PPTC  \\\n",
       "Histology.Priors                                                  \n",
       "BCP-ALL                   37     37        37                37   \n",
       "Brain                     48     48        48                48   \n",
       "Carcinoma                  3      3         3                 3   \n",
       "ETP-ALL                    6      6         6                 6   \n",
       "Ewing Sarcoma              9      9         9                 9   \n",
       "Extracranial Rhabdoid      3      3         3                 3   \n",
       "MLL-ALL                   10     10        10                10   \n",
       "Neuroblastoma             33     33        33                33   \n",
       "Osteosarcoma              30     30        30                30   \n",
       "Other Sarcoma              3      3         3                 3   \n",
       "Ph+/like-ALL              18     18        18                18   \n",
       "Renal                     13     13        13                13   \n",
       "Rhabdomyosarcoma          12     12        12                12   \n",
       "T-ALL                     19     19        19                19   \n",
       "\n",
       "                       RNA.Part.of.PPTC  Have.snp.file  snp.array.filename  \\\n",
       "Histology.Priors                                                             \n",
       "BCP-ALL                              37             37                  36   \n",
       "Brain                                48             48                  44   \n",
       "Carcinoma                             3              3                   3   \n",
       "ETP-ALL                               6              6                   6   \n",
       "Ewing Sarcoma                         9              9                   9   \n",
       "Extracranial Rhabdoid                 3              3                   3   \n",
       "MLL-ALL                              10             10                  10   \n",
       "Neuroblastoma                        33             33                  33   \n",
       "Osteosarcoma                         30             30                  30   \n",
       "Other Sarcoma                         3              3                   3   \n",
       "Ph+/like-ALL                         18             18                  18   \n",
       "Renal                                13             13                  13   \n",
       "Rhabdomyosarcoma                     12             12                  12   \n",
       "T-ALL                                19             19                  19   \n",
       "\n",
       "                       snp.array.sample.ID  Have.maf  WES.human.bam.filename  \\\n",
       "Histology.Priors                                                               \n",
       "BCP-ALL                                 36        37                      37   \n",
       "Brain                                   44        48                      32   \n",
       "Carcinoma                                3         3                       3   \n",
       "ETP-ALL                                  6         6                       6   \n",
       "Ewing Sarcoma                            9         9                       9   \n",
       "Extracranial Rhabdoid                    3         3                       3   \n",
       "MLL-ALL                                 10        10                      10   \n",
       "Neuroblastoma                           33        33                      33   \n",
       "Osteosarcoma                            30        30                      28   \n",
       "Other Sarcoma                            3         3                       3   \n",
       "Ph+/like-ALL                            18        18                      18   \n",
       "Renal                                   13        13                      13   \n",
       "Rhabdomyosarcoma                        12        12                      12   \n",
       "T-ALL                                   19        19                      19   \n",
       "\n",
       "                       ...  Histology.Oncoprints2  STR.OK  Model.Type  \\\n",
       "Histology.Priors       ...                                              \n",
       "BCP-ALL                ...                     37      37          37   \n",
       "Brain                  ...                     48      48          48   \n",
       "Carcinoma              ...                      3       3           3   \n",
       "ETP-ALL                ...                      6       6           6   \n",
       "Ewing Sarcoma          ...                      9       9           9   \n",
       "Extracranial Rhabdoid  ...                      3       3           3   \n",
       "MLL-ALL                ...                     10      10          10   \n",
       "Neuroblastoma          ...                     33      33          33   \n",
       "Osteosarcoma           ...                     30      30          30   \n",
       "Other Sarcoma          ...                      3       3           3   \n",
       "Ph+/like-ALL           ...                     18      18          18   \n",
       "Renal                  ...                     13      13          13   \n",
       "Rhabdomyosarcoma       ...                     12      12          12   \n",
       "T-ALL                  ...                     19      19          19   \n",
       "\n",
       "                       Histology  Histology.Detailed  PI  Sex  Age  \\\n",
       "Histology.Priors                                                     \n",
       "BCP-ALL                       37                  37  37   37   37   \n",
       "Brain                         48                  48  48   48   42   \n",
       "Carcinoma                      3                   3   3    3    3   \n",
       "ETP-ALL                        6                   6   6    6    6   \n",
       "Ewing Sarcoma                  9                   9   9    9    9   \n",
       "Extracranial Rhabdoid          3                   3   3    3    3   \n",
       "MLL-ALL                       10                  10  10   10   10   \n",
       "Neuroblastoma                 33                  33  33   33   28   \n",
       "Osteosarcoma                  30                  30  30   30   25   \n",
       "Other Sarcoma                  3                   3   3    3    1   \n",
       "Ph+/like-ALL                  18                  18  18   18   18   \n",
       "Renal                         13                  13  13   13   11   \n",
       "Rhabdomyosarcoma              12                  12  12   12   11   \n",
       "T-ALL                         19                  19  19   19   19   \n",
       "\n",
       "                       Reported_Ethnicity  Inferred_Ethnicity  \n",
       "Histology.Priors                                               \n",
       "BCP-ALL                                36                  36  \n",
       "Brain                                  44                  44  \n",
       "Carcinoma                               3                   3  \n",
       "ETP-ALL                                 6                   6  \n",
       "Ewing Sarcoma                           9                   9  \n",
       "Extracranial Rhabdoid                   3                   3  \n",
       "MLL-ALL                                10                  10  \n",
       "Neuroblastoma                          33                  33  \n",
       "Osteosarcoma                           30                  30  \n",
       "Other Sarcoma                           3                   3  \n",
       "Ph+/like-ALL                           18                  18  \n",
       "Renal                                  13                  13  \n",
       "Rhabdomyosarcoma                       12                  12  \n",
       "T-ALL                                  19                  19  \n",
       "\n",
       "[14 rows x 31 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx_clinical.groupby('Histology.Priors').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('ASPS', 'Other Sarcoma') 2\n",
      "('ATRT', 'Brain') 6\n",
      "('Astrocytoma', 'Brain') 4\n",
      "('BCP-ALL', 'BCP-ALL') 33\n",
      "('CNS EFT-CIC', 'Brain') 2\n",
      "('CNS embryonal NOS', 'Brain') 1\n",
      "('CNS germinoma', 'Brain') 1\n",
      "('Clear Cell Sarcoma', 'Other Sarcoma') 1\n",
      "('Colon Carcinoma', 'Carcinoma') 1\n",
      "('DIPG', 'Brain') 2\n",
      "('ETMR', 'Brain') 3\n",
      "('ETP-ALL', 'ETP-ALL') 6\n",
      "('Ependymoblastoma', 'Brain') 1\n",
      "('Ependymoma', 'Brain') 6\n",
      "('Ewing Sarcoma', 'Ewing Sarcoma') 9\n",
      "('Extracranial Rhabdoid', 'Extracranial Rhabdoid') 3\n",
      "('Fusion+ RMS', 'Rhabdomyosarcoma') 6\n",
      "('Fusion- RMS', 'Rhabdomyosarcoma') 6\n",
      "('Glioblastoma', 'Brain') 4\n",
      "('Hepatoblastoma', 'Renal') 1\n",
      "('MLL-ALL', 'MLL-ALL') 10\n",
      "('Medulloblastoma', 'Brain') 18\n",
      "('Neuroblastoma', 'Neuroblastoma') 33\n",
      "('Osteosarcoma', 'Osteosarcoma') 30\n",
      "('Ph+-ALL', 'Ph+/like-ALL') 3\n",
      "('Ph-likeALL', 'BCP-ALL') 4\n",
      "('Ph-likeALL', 'Ph+/like-ALL') 15\n",
      "('Small Cell Carcinoma', 'Carcinoma') 2\n",
      "('T-ALL', 'T-ALL') 19\n",
      "('Wilms', 'Renal') 12\n"
     ]
    }
   ],
   "source": [
    "for i, rows in pdx_clinical.groupby(['Histology.Detailed', 'Histology.Priors']):\n",
    "    print(i, len(rows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tissue(d):\n",
    "    \"\"\"\n",
    "    We found that including a tissue specific prior improves \n",
    "    identification of differentially expressed genes across PDXs.\n",
    "    Here we match specific diseases to broad tissue categories.\n",
    "    \"\"\"\n",
    "    if d in ['ATRT', \n",
    "             'Astrocytoma', \n",
    "             'CNS germinoma', \n",
    "             'DIPG', \n",
    "             'Ependymoma',\n",
    "             'ETMR',\n",
    "             'Glioblastoma', \n",
    "             'High-grade glioma',\n",
    "             'Medulloblastoma', \n",
    "             'PNET',\n",
    "             'Anaplastic Rhabdoid Meningioma',\n",
    "             'CNS EFT-CIC',\n",
    "             'CNS embryonal NOS',\n",
    "             'Ependymoblastoma']:\n",
    "        return \"brain\"\n",
    "    \n",
    "    elif d in ['ETP-ALL',\n",
    "               'T-ALL']:\n",
    "        return 't-cell ALL'\n",
    "    \n",
    "    elif d in ['BCP-ALL',\n",
    "               'MLL-ALL',\n",
    "               'Ph+-ALL',\n",
    "               'Ph-likeALL']:\n",
    "        return 'b-cell ALL'\n",
    "    \n",
    "    elif d in ['Ewing Sarcoma', 'Osteosarcoma']:\n",
    "        return 'bone'\n",
    "    \n",
    "    elif d in ['ASPS',\n",
    "               'Other Sarcoma',\n",
    "               'Rhabdoid',\n",
    "               'Rhabdomyosarcoma',\n",
    "               'Alveolar Rhabdomyosarcoma',\n",
    "               'Embryonal Rhabdomyosarcoma',\n",
    "               'Extracranial Rhabdoid',\n",
    "               'Fusion+ RMS',\n",
    "               'Fusion- RMS',\n",
    "               'Clear Cell Sarcoma']:\n",
    "        return 'soft sarcoma'\n",
    "    \n",
    "    elif d in ['Wilms', 'Other Renal']:\n",
    "        return 'renal'\n",
    "    \n",
    "    elif d == 'Neuroblastoma':\n",
    "        return 'neuroblastoma'\n",
    "    \n",
    "    elif d == 'Hepatoblastoma':\n",
    "        return 'hepatoblastoma'\n",
    "    \n",
    "    elif d in ['Small Cell Carcinoma', 'Colon Carcinoma']:\n",
    "        return 'carcinoma'\n",
    "    \n",
    "    else:\n",
    "        raise ValueError(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Match meta data and expression to each PDX. \n",
    "tmps = []\n",
    "for p in exp.columns:\n",
    "    if p in pdxs:\n",
    "        tmp = pd.DataFrame(columns=['sample', \n",
    "                                    'tissue', \n",
    "                                    'disease', \n",
    "                                    'gene', \n",
    "                                    'gene_id', \n",
    "                                    'expression'],\n",
    "                           index=range(len(exp.index)))\n",
    "        \n",
    "        disease = pdx_clinical.loc[pdx_clinical['Model'] == p, 'Histology.Detailed'].item()\n",
    "        \n",
    "        tissue = get_tissue(disease)\n",
    "        \n",
    "        tmp['sample'] = p\n",
    "        tmp['tissue'] = tissue\n",
    "        tmp['disease'] = disease\n",
    "        tmp['gene'] = exp.index.values\n",
    "        tmp['gene_id'] = range(1, len(exp.index.values) + 1)\n",
    "        tmp['expression'] = exp.loc[exp.index, p].values\n",
    "        \n",
    "    else:\n",
    "        raise ValuError()\n",
    "        \n",
    "    tmps.append(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge all of these features into one large dataframe\n",
    "data = pd.concat(tmps,\n",
    "                 axis=0,\n",
    "                 verify_integrity=True,\n",
    "                 ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that the mean is centered\n",
    "for i, row in data.groupby('gene'):\n",
    "    assert np.isclose(np.mean(row['expression'].values), [0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stan only takes integer values, so match disease and tissue names \n",
    "# to an integer identifier.\n",
    "disease_map = dict((d, i) for d, i in zip(data['disease'].unique(), range(1, len(data['disease'].unique()) + 1)))\n",
    "disease_map\n",
    "\n",
    "tissue_map = dict((d, i) for d, i in zip(data['tissue'].unique(), range(1, len(data['tissue'].unique()) + 1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add disease and tissue identifiers to dataframe\n",
    "data['disease_id'] = data['disease'].map(disease_map)\n",
    "data['tissue_id'] = data['tissue'].map(tissue_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instead of using the tissue_id, stan uses \n",
    "# the disease to tissue map.\n",
    "disease_tissue_map = []\n",
    "for disease in sorted(data['disease_id'].unique()):\n",
    "    tissue = data.loc[data['disease_id'] == disease, 'tissue_id'].unique()[0]\n",
    "    disease_tissue_map.append(tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "stan_d = {'N': data.shape[0],\n",
    "          'G': len(data['gene_id'].unique()),\n",
    "          'T': len(data['tissue_id'].unique()),\n",
    "          'D': len(data['disease_id'].unique()),\n",
    "          'genes': data['gene_id'].values,\n",
    "          'tissues': disease_tissue_map,\n",
    "          'diseases': data['disease_id'].values,\n",
    "          'y': data['expression'].values}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the Stan data dictionary to disk\n",
    "pth = os.path.join(data_dir, 'stan-data-v8-2019-07-30.dump')\n",
    "pystan.stan_rdump(stan_d, pth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the mapping to the original disease, tissue, and gene names\n",
    "pth = os.path.join(data_dir, 'stan-data-v8-map-2019-07-30.pkl')\n",
    "inv_disease_map = {v: k for k, v in disease_map.items()}\n",
    "inv_tissue_map = {v: k for k, v in tissue_map.items()}\n",
    "\n",
    "gene_mapper = {}\n",
    "for gene, _id in zip(data['gene'].unique(), data['gene_id'].unique()):\n",
    "    gene_mapper[_id] = gene\n",
    "\n",
    "maps = {'tissue': inv_tissue_map,\n",
    "        'disease': inv_disease_map, \n",
    "        'gene': gene_mapper}\n",
    "\n",
    "with open(pth, 'wb') as f:\n",
    "    pickle.dump(maps, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 'BCP-ALL',\n",
       " 2: 'MLL-ALL',\n",
       " 3: 'Ph+-ALL',\n",
       " 4: 'T-ALL',\n",
       " 5: 'Ph-likeALL',\n",
       " 6: 'ASPS',\n",
       " 7: 'CNS embryonal NOS',\n",
       " 8: 'ATRT',\n",
       " 9: 'Ewing Sarcoma',\n",
       " 10: 'Neuroblastoma',\n",
       " 11: 'ETP-ALL',\n",
       " 12: 'Extracranial Rhabdoid',\n",
       " 13: 'ETMR',\n",
       " 14: 'DIPG',\n",
       " 15: 'Glioblastoma',\n",
       " 16: 'Ependymoma',\n",
       " 17: 'Ependymoblastoma',\n",
       " 18: 'CNS EFT-CIC',\n",
       " 19: 'Astrocytoma',\n",
       " 20: 'CNS germinoma',\n",
       " 21: 'Medulloblastoma',\n",
       " 22: 'Fusion- RMS',\n",
       " 23: 'Wilms',\n",
       " 24: 'Fusion+ RMS',\n",
       " 25: 'Small Cell Carcinoma',\n",
       " 26: 'Colon Carcinoma',\n",
       " 27: 'Clear Cell Sarcoma',\n",
       " 28: 'Hepatoblastoma',\n",
       " 29: 'Osteosarcoma'}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inv_disease_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that the gene and gene_id values are the same for \n",
    "# all of the samples\n",
    "last = None\n",
    "for i, row in data.groupby('sample'):\n",
    "    if last is None:\n",
    "        last = row\n",
    "        continue\n",
    "        \n",
    "    assert (row['gene'].values == last['gene'].values).all()\n",
    "    assert (row['gene_id'].values == last['gene_id'].values).all()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
